### PR TITLE
feat(core): provide git branch to template strings

### LINKED ISSUE #2086: [FEATURE]: Make current git branch available as variable
## Feature Request
Make the current git branch available as a pre-defined garden variable.

### Background / Motivation
We would like to use this as part of our defaultNamespace value. Currently we either have to deploy by manually specifying the branch name with --env or ensure the branch name is available as an evironment variable. 

### What should the user be able to do?
Set, using say local.git-branch, the value of the current branch anywhere in the garden project

### Why do they want to do this? What problem does it solve?
It creates an easy and automated way to work with multiple feature branches by linking them to garden environments.

### How important is this feature for you/your team?

🌹 It’s a nice to have, but nice things are nice 🙂

### PR DESCRIPTION
**Pull request by thsig**
**What this PR does / why we need it**:

The current Git branch can now be used in template strings under `${git.branch}`.

**Which issue(s) this PR fixes**:

Closes #2086.

### CODE REVIEW

**Comment by thsig** [comment]
Ah, looks like I used an option that's not supported by the Git version used in the test suite. Will have a look.

**Comment by edvald** [comment]
I wonder if it makes more sense to make this `${git.branch}` instead? Or `${local.git.branch}` perhaps.

**Comment by thsig** [comment]
I've changed this template variable to `${git.branch}` now, so I think we should be good here.

File: core/src/config/config-context.ts [hunk 1]
```diff
@@ -340,16 +364,6 @@ export interface ProjectConfigContextParams {
...
 export class ProjectConfigContext extends DefaultEnvironmentContext {
-  @schema(
-    LocalContext.getSchema().description(
-      "Context variables that are specific to the currently running environment/machine."
-    )
-  )
-  public local: LocalContext
-
-  @schema(ProjectContext.getSchema().description("Information about the Garden project."))
-  public project: ProjectContext
```
Comment (edvald) [inline core/src/config/config-context.ts:351]: I'm assuming this is unintentional?

Comment (thsig) [reply core/src/config/config-context.ts:351]: It's intentional, actually—`ProjectConfigContext` extends `DefaultEnvironmentContext`, which has the `local` and `project` keys, so these declarations were essentially duplicates.

The relevant tests pass (and the generated docs haven't changed when it comes to these variables).

**Comment by thsig** [comment]
Tests are still failing in CI (but pass locally). Maybe the git versions are different.

File: core/src/config/config-context.ts (hunk 1 above)
Comment (edvald) [reply core/src/config/config-context.ts:351]: Right, got it.

File: core/src/vcs/git.ts [hunk 2]
```diff
@@ -417,4 +409,30 @@ export class GitHandler extends VcsHandler {
...
+      return (await git("rev-parse", "--abbrev-ref", "HEAD"))[0]
+    } catch (err) {
+      if (err.exitCode === 128) {
+        try {
+          // If this doesn't throw, then we're in a repo with no commits.
+          await git("rev-list", "-n", "1", "--all")
+          return undefined
+        } catch (notInRepoError) {
+          // Throw nice error when we detect that we're not in a repo root
+          throw new RuntimeError(notInRepoRootErrorMessage(path), { path })
```
Comment (edvald) [inline core/src/vcs/git.ts:425]: We need a more specific error check here (the git command could be failing in any number of ways other than not being in a repo root). That should bring to light what's actually failing in the tests.

Comment (edvald) [inline core/src/vcs/git.ts:425]: (Oops, meant to edit and not delete what I previously wrote)

It seems we also need a more specific error check for this nested git command. Or perhaps we should generally fail silently here, and just document that `${git.branch}` may be undefined in certain cases? We might for example be on a detached HEAD (e.g. rebasing).

**Comment by thsig** [comment]
I think this should be good now. Are the comments etc. clear enough?

**Comment by thsig** [comment]
Are there any outstanding comments on this one?

**Comment by edvald** [comment]
One thought: If running in watch mode, the branch won't update unless Garden configuration is changed. I wonder if it makes sense to at least document that? The alternative would be to watch something in the `.git` folder but I'm not sure we want to get into such things now.

Also, it's not defined in documentation what to expect when using remote sources. I think it makes sense to always use the project repo branch (although I suppose it is kind of debatable), but we should at least clarify that in docs.

**Comment by thsig** [comment]
> One thought: If running in watch mode, the branch won't update unless Garden configuration is changed. I wonder if it makes sense to at least document that?

Yeah, agreed. I'll add a description of that to the docstring.

> The alternative would be to watch something in the .git folder but I'm not sure we want to get into such things now.

Aye. That was my thought initially, but I think we should leave that for another PR if that feature request comes up from users.

> Also, it's not defined in documentation what to expect when using remote sources. I think it makes sense to always use the project repo branch (although I suppose it is kind of debatable), but we should at least clarify that in docs.

Good point. I'll describe the current behavior more fully in the docstring.

**Comment by thsig** [comment]
I've updated the docstring.

**Review by edvald — APPROVED** [review]

**Pull request merged**